int-set/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+mod bitpage;
+mod bitset;
+
+use bitset::BitSet;
+use std::ops::RangeInclusive;
+
+/// A fast & efficient unsigned integer (u32) bit set which is invertible.
+#[derive(Clone, Debug)]
+pub enum IntSet<T> {
+    /// Records a set of integers which are members of the set.
+    Standard(BitSet<T>),
+
+    /// Records the set of integers which are not members of the set.
+    Inverted(BitSet<T>),
+}
+
+impl<T: Into<u32> + Copy + Default> IntSet<T> {
+    /// Create a new empty set.
+    pub fn default() -> IntSet<T> {
+        IntSet::Standard(BitSet::<T>::default())
+    }
+
+    /// Create a new set which contains all integers.
+    pub fn all() -> IntSet<T> {
+        IntSet::Inverted(BitSet::<T>::default())
+    }
+
+    /// Return the inverted version of this set.
+    pub fn inverted(self) -> IntSet<T> {
+        match self {
+            IntSet::<T>::Standard(s) => IntSet::<T>::Inverted(s),
+            IntSet::<T>::Inverted(s) => IntSet::<T>::Standard(s),
+        }
+    }
+
+    /// Return a new version of this set with all members removed.
```

Comment:
from `HashSet::clear`:

```suggestion
    /// Clears the set, removing all values.
```

(By convention clear is a mutating method on self, and does not return anything.)

Reply:
Because to properly clear the set we have to change the enum type I believe this requires returning a new IntSet (similar to the inverted() method).

Reply:
You should be able to do:

```rust
pub fn clear(&mut self) {
    self.clear_internal_set();
    *self.0 = match self.0 {
        Membership::Inclusive(s) => IntSet(Membership::Inclusive(s)),
        Membership::Exclusive(s) => IntSet(Membership::Inclusive(s)),
    };
}
```

if you want.

Reply:
That doesn't work unfortunately, I've played around with it a bit and the closest I could get to making it work is this:

pub fn clear(&mut self) {
        self.clear_internal_set();
        self.0 = match &mut self.0 {
            Membership::Inclusive(s) => Membership::Inclusive(*s),
            Membership::Exclusive(s) => Membership::Inclusive(*s),
        };
    }
    
 But that still has problems:
 
 error[E0507]: cannot move out of `*s` which is behind a mutable reference

Reply:
Ah okay yea, that's a funny little hiccup.

There are two basic options:

- if we want to forget about reusing the storage we can just replace the bitset, like:
   ```rust
    *s = Membership::Inclusive(BitSet::empty());
   ```
- and if we want to preserve the existing set, we can do a little dance to get around borrowck:
```rust
/// Return a new version of this set with all members removed.
    pub fn clear(&mut self) {
        let reuse_storage = match &mut self.0 {
            // if we're inclusive, we just clear the storage
            Membership::Inclusive(s) => {
                s.clear();
                return;
            }
            // otherwise take the existing storage to reuse in a new
            // inclusive set:
            Membership::Exclusive(s) => std::mem::replace(s, BitSet::empty()),
        };
        // reuse the now empty storage and mark us as inclusive
        self.0 = Membership::Inclusive(reuse_storage);
    }
```

Reply:
I was hoping to keep the allocated memory in the set around, so looks like I'll go with the second version.